proposal/color-4-new-spaces-js.d.ts.md

```diff
@@ -0,0 +1,864 @@
+
+* Let `h` be the result of [parsing a channel value] with value `options.h`.
+  
+* If `options.alpha` is not set, let `alpha` be `1`. Otherwise, let `alpha` be
+    the result of [parsing a channel value] with value `options.alpha`.
+
+* If `options.space` equals `lch`, set [`internal`] to the result of
+  [`lch(lightness a b / alpha)`].
+
+* Otherwise, if `options.space` equals `oklch`, set [`internal`] to the result
+  of [`oklch(lightness a b / alpha)`].
+
+```ts
+constructor(options: {
+  lightness: number | null;
+  chroma: number | null;
+  hue: number | null;
+  alpha?: number | null;
+  space: ColorSpaceLCH;
+});
+```
+
+#### Predefined RGB Channel Constructor
+
+Create a new SassColor in a color space with RGB channels -- `srgb`,
+`srgb-linear`, `display-p3`, `a98-rgb`, and `prophoto-rgb`. `rgb` is not
+supported with this constructor, as the legacy `rgb` color does not support
+[missing components].
+
+* Let `red` be the result of [parsing a channel value] with value `options.red`.
+
+* Let `green` be the result of [parsing a channel value] with value
+  `options.green`.
+
+* Let `blue` be the result of [parsing a channel value] with value
+  `options.blue`.
+  
+* If `options.alpha` is not set, let `alpha` be `1`. Otherwise, let `alpha` be
+    the result of [parsing a channel value] with value `options.alpha`.
+
+* Let `space` be the unquoted string value of `options.space`.
+
+* Set [`internal`] to the result of [`color(space red green blue / alpha)`].
+
+[`color(space red green blue / alpha)`]: ./color-4-new-spaces.md#color-1
+
+```ts
+constructor(options: {
+  red: number | null;
+  green: number | null;
+  blue: number | null;
+  alpha?: number | null;
+  space: Omit<ColorSpaceRGB, 'rgb'>;
+});
+```
+
+#### XYZ Channel Constructor
+
+Create a new SassColor in a color space with XYZ channels -- `xyz`, `xyz-d50`,
+and `xyz-d65`.
+
+* Let `x` be the result of [parsing a channel value] with value `options.x`.
+
+* Let `y` be the result of [parsing a channel value] with value `options.y`.
+
+* Let `z` be the result of [parsing a channel value] with value `options.z`.
+  
+* If `options.alpha` is not set, let `alpha` be `1`. Otherwise, let `alpha` be
+    the result of [parsing a channel value] with value `options.alpha`.
+
+* Let `space` be the unquoted string value of `options.space`.
+
+* Set [`internal`] to the result of [`color(space x y z / alpha)`].
+
+```ts
+constructor(options: {
+  x: number | null;
+  y: number | null;
+  z: number | null;
+  alpha?: number | null;
+  space: ColorSpaceXYZ;
+});
+```
+
+#### Legacy Color Constructors with `space`
+
+While the [legacy color space] constructors do not require a space, replace the
+[existing types] with these constuctor overloads for forward compatibility.
+
+[existing types]: ../spec/js-api/value/color.d.ts.md#constructor
+
+```ts
+constructor(options: {
+  red: number;
+  green: number;
+  blue: number;
```

Comment:
Shouldn't we allow `null` here so users can set missing channels for these spaces?

Reply:
Good question- I think we need to do some more explicit handling here. The [New Color Spaces Sass Proposal](https://github.com/sass/sass/blob/main/proposal/color-4-new-spaces.md#rgb-and-rgba) explicitly says for RGB and HSL `Missing channels are not allowed in legacy syntax.`. I think this is where I was getting confused elsewhere with legacy colors not supporting missing channels, but it's actually that the legacy Sass constructors for RGB and HSL don't allow missing colors, but they also have new Sass constructors that do.

I think I have a plan, but want to double check-

1. Replace the [JS API color constructor](https://github.com/sass/sass/blob/main/spec/js-api/value/color.d.ts.md#constructor) instead of referring to it
2. For RGB, if space is not set, round channels to an integer to preserve past functionality (and emit deprecation warning? Maybe only if a non-integer is provided?)
3. Set `internal` to the result of `rgb(red green blue / alpha)` instead of `rgb(red, green, blue, alpha)` (and similar for hsl and hwb.) With these constructors, `none` is allowed.
4. Because we're switching all construction to use the modern space-separated syntax, we can allow `null` even if `space` is not set.

@nex3  Is that all correct?

Reply:
I implemented this in eac2876322245614b5e3d2e7bd8dceb9a9a483dd, and decided not to emit a deprecation in step 2.

Reply:
This all sounds good, but I wouldn't round in step 2.

Reply:
I removed the rounding in https://github.com/sass/sass/pull/3665/commits/00b80f55b1b2af0157716a81c6a1cbbe02fd2769. 

I also realized that I could emit the null-alpha deprecation if alpha is null and space is not defined, so I did that in that commit for RGB, and in HSL and HWB in https://github.com/sass/sass/pull/3665/commits/d0fb9e0f0b3174c7907a57f42eff2f43a20ded6f.